torchao/prototype/smoothquant/api.py

```diff
@@ -5,186 +5,119 @@
-        wei_scales: The weight scales for the layer. Acquired from the layer's observer if None.
         set_inductor_config: if True, adjusts `torchinductor` settings to recommended values.
     """
 
+    base_config: AOBaseConfig
+    step: SmoothQuantStep
+    alpha: Optional[float] = 0.5
     smoothing_factor: Optional[torch.Tensor] = None
-    act_scales: Optional[torch.Tensor] = None
-    wei_scales: Optional[torch.Tensor] = None
     set_inductor_config: bool = True
 
+    def __post_init__(self):
+        self.step = self.step.lower() if isinstance(self.step, str) else self.step.value
+        all_step_values = [s.value for s in SmoothQuantStep]
+        if self.step not in all_step_values:
+            raise ValueError(f"{self.step} is not one of {all_step_values}")
+
 
 @register_quantize_module_handler(SmoothQuantConfig)
 def _smooth_quant_transform(
     module: torch.nn.Module,
     config: SmoothQuantConfig,
-):
-    smoothing_factor = config.smoothing_factor
-    act_scales = config.act_scales
-    wei_scales = config.wei_scales
+) -> torch.nn.Module:
+    step = config.step
+    observed_linear = None
+    base_config = config.base_config
+
+    if step == SmoothQuantStep.PREPARE:
+        observer = SmoothQuantObserver(
+            weight=module.weight,
+            bias=module.bias,
+            base_config=base_config,
+            alpha=config.alpha,
+        )
+        return SmoothQuantObservedLinear.from_float(module, observer)
+
+    elif step == SmoothQuantStep.PREPARE_FOR_LOADING:
+        # loading from pre-quantized checkpoint
+        observer = SmoothQuantObserver(
+            weight=module.weight,
+            bias=module.bias,
+            base_config=base_config,
+            alpha=config.alpha,
+        )
+        observed_linear = SmoothQuantObservedLinear.from_float(module, observer)
+        example_input = torch.randn(
+            (1, module.weight.shape[1]),
+            device=module.weight.device,
+            dtype=module.weight.dtype,
+        )
+        observed_linear(example_input)
+
+    elif step == SmoothQuantStep.CONVERT:
+        if not isinstance(module, SmoothQuantObservedLinear):
+            print(
+                f"convert: module is not SmoothQuantObservedLinear, skipping: {type(module)}"
+            )
+            return module
+        observed_linear = module
+    else:
+        raise ValueError(f"Unexpected step: {step}")
+
+    # Convert observed linear to quantized linear
     if config.set_inductor_config:
         torchao.quantization.utils.recommended_inductor_config_setter()
-    observed_linear = module
 
+    # Get quantization parameters
+    smoothing_factor = (
+        config.smoothing_factor
+        if config.smoothing_factor is not None
+        else observed_linear.obs.calculate_qparams()
+    )
+    weight = observed_linear.weight * smoothing_factor
+
+    # Create new linear layer
     linear = torch.nn.Linear(
```

Comment:
nit: one trick we can have when creating these linear weights is to create them in meta device: https://github.com/vllm-project/vllm/blob/c86af22f31838ee654c856279ac5110ae3fdb2cc/vllm/model_executor/layers/quantization/torchao.py#L159 to save memory I think, similar for awq